// @fluentui/react-charts/lib-commonjs/components/FunnelChart/useFunnelChartStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    funnelClassNames: function() {
        return funnelClassNames;
    },
    useFunnelChartStyles: function() {
        return useFunnelChartStyles;
    }
});
const _react = require("@griffel/react");
const funnelClassNames = {
    root: 'fui-funnel__root',
    chart: 'fui-funnel__chart',
    text: 'fui-funnel__text',
    calloutContentRoot: 'fui-funnel__callout-content-root'
};
/**
 * Base Styles
 */ const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        Bt984gj: "f122n59",
        a9b677: "fly5x3f",
        Bqenvij: "f1l02sjl",
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508",
        Bhrd7zp: "figsok6",
        qhf8xq: "f10pi13n"
    },
    chart: {
        mc9l5x: "ftgm304",
        a9b677: "fly5x3f",
        Bqenvij: "f1l02sjl"
    },
    text: {
        Bkecrkj: "f1aehjj5",
        Bhrd7zp: "fl43uef",
        Bkfmm31: "fhuob2q",
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508"
    },
    calloutContentRoot: {
        B2u0y6b: "f15g8at8"
    }
}, {
    d: [
        ".f122n59{align-items:center;}",
        ".fly5x3f{width:100%;}",
        ".f1l02sjl{height:100%;}",
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fkhj508{font-size:var(--fontSizeBase300);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".f10pi13n{position:relative;}",
        ".ftgm304{display:block;}",
        ".f1aehjj5{pointer-events:none;}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fhuob2q{fill:var(--colorNeutralForeground1);}",
        ".f15g8at8{max-width:238px;}"
    ]
});
const useFunnelChartStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2, _props_styles3;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(funnelClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        chart: (0, _react.mergeClasses)(funnelClassNames.chart, baseStyles.chart, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.chart),
        text: (0, _react.mergeClasses)(funnelClassNames.text, baseStyles.text, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.text),
        calloutContentRoot: (0, _react.mergeClasses)(baseStyles.calloutContentRoot, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.calloutContentRoot)
    };
};
