// @fluentui/react-charts/lib-commonjs/components/FunnelChart/useFunnelChartStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    funnelClassNames: function() {
        return funnelClassNames;
    },
    useFunnelChartStyles: function() {
        return useFunnelChartStyles;
    }
});
const _react = require("@griffel/react");
const _reacttheme = require("@fluentui/react-theme");
const funnelClassNames = {
    root: 'fui-funnel__root',
    chart: 'fui-funnel__chart',
    text: 'fui-funnel__text',
    calloutContentRoot: 'fui-funnel__callout-content-root'
};
/**
 * Base Styles
 */ const useStyles = (0, _react.makeStyles)({
    root: {
        alignItems: 'center',
        width: '100%',
        height: '100%',
        fontFamily: _reacttheme.tokens.fontFamilyBase,
        fontSize: _reacttheme.tokens.fontSizeBase300,
        fontWeight: _reacttheme.tokens.fontWeightRegular,
        position: 'relative'
    },
    chart: {
        display: 'block',
        width: '100%',
        height: '100%'
    },
    text: {
        pointerEvents: 'none',
        fontWeight: _reacttheme.tokens.fontWeightSemibold,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        fontFamily: _reacttheme.tokens.fontFamilyBase,
        fontSize: _reacttheme.tokens.fontSizeBase300
    },
    calloutContentRoot: {
        maxWidth: '238px'
    }
});
const useFunnelChartStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2, _props_styles3;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(funnelClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        chart: (0, _react.mergeClasses)(funnelClassNames.chart, baseStyles.chart, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.chart),
        text: (0, _react.mergeClasses)(funnelClassNames.text, baseStyles.text, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.text),
        calloutContentRoot: (0, _react.mergeClasses)(baseStyles.calloutContentRoot, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.calloutContentRoot)
    };
};
