// @fluentui/react-charts/lib-commonjs/components/DonutChart/Pie/usePieStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    donutPieClassNames: function() {
        return donutPieClassNames;
    },
    usePieStyles: function() {
        return usePieStyles;
    }
});
const _react = require("@griffel/react");
const donutPieClassNames = {
    root: 'fui-donut-pie__root',
    insideDonutString: 'fui-donut-pie__insideDonutString'
};
/**
 * Base Styles
 */ const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {},
    insideDonutString: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fojgt09",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fcen8rp",
        Bkfmm31: "fhuob2q",
        Bvjb7m6: "fhv2zbx"
    }
}, {
    d: [
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fojgt09{font-size:var(--fontSizeHero700);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fcen8rp{line-height:var(--lineHeightHero700);}",
        ".fhuob2q{fill:var(--colorNeutralForeground1);}",
        ".fhv2zbx{forced-color-adjust:auto;}"
    ]
});
const usePieStyles = (props)=>{
    var _props_styles, _props_styles1;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(donutPieClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        insideDonutString: (0, _react.mergeClasses)(donutPieClassNames.insideDonutString, baseStyles.insideDonutString, className, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.insideDonutString)
    };
};
