// @fluentui/react-charts/lib-commonjs/components/DonutChart/Pie/usePieStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    donutPieClassNames: function() {
        return donutPieClassNames;
    },
    usePieStyles: function() {
        return usePieStyles;
    }
});
const _reacttheme = require("@fluentui/react-theme");
const _react = require("@griffel/react");
const donutPieClassNames = {
    root: 'fui-donut-pie__root',
    insideDonutString: 'fui-donut-pie__insideDonutString'
};
/**
 * Base Styles
 */ const useStyles = (0, _react.makeStyles)({
    root: {},
    insideDonutString: {
        ..._reacttheme.typographyStyles.title2,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    }
});
const usePieStyles = (props)=>{
    var _props_styles, _props_styles1;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(donutPieClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        insideDonutString: (0, _react.mergeClasses)(donutPieClassNames.insideDonutString, baseStyles.insideDonutString, className, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.insideDonutString)
    };
};
