// @fluentui/react-charts/lib/components/HorizontalBarChart/useHorizontalBarChartStyles.styles.js
'use client';

import { __styles, mergeClasses, shorthands } from '@griffel/react';
import { tokens, typographyStyles } from '@fluentui/react-theme';
import { HorizontalBarChartVariant } from './index';
import { HighContrastSelector } from '../../utilities/index';
/**
 * @internal
 */
export const hbcClassNames = {
  root: 'fui-hbc__root',
  items: 'fui-hbc__items',
  chart: 'fui-hbc__chart',
  chartTitle: 'fui-hbc__chartTitle',
  barWrapper: 'fui-hbc__barWrapper',
  chartTitleLeft: 'fui-hbc__chartTitleLeft',
  chartTitleRight: 'fui-hbc__chartTitleRight',
  chartDataTextDenominator: 'fui-hbc__textDenom',
  benchmarkContainer: 'fui-hbc__benchmark',
  triangle: 'fui-hbc__triangle',
  barLabel: 'fui-hbc__barLabel',
  chartWrapper: 'fui-hbc__chartWrapper',
  legendContainer: 'fui-hbc__legendContainer'
};
/**
 * Base Styles
 */
const useStyles = /*#__PURE__*/__styles({
  root: {
    mc9l5x: "f22iagw",
    Beiy3e4: "f1vx9l62",
    a9b677: "fly5x3f"
  },
  items10pMargin: {
    jrapky: "f1rfdi54"
  },
  items16pMargin: {
    jrapky: "fbxl46w"
  },
  chart: {
    a9b677: "fly5x3f",
    Bqenvij: "fvblgha",
    mc9l5x: "ftgm304",
    B68tc82: 0,
    Bmxbyg5: 0,
    Bpg54ce: "f1gl81tg"
  },
  barWrapper: {},
  chartTitle: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "figsok6",
    Bg96gwp: "fwrc4pm",
    mc9l5x: "f22iagw",
    Brf1p80: "f1869bpl"
  },
  chartTitleLeft: {
    ygn44y: "f1cmbuwj",
    B68tc82: 0,
    Bmxbyg5: 0,
    Bpg54ce: "f1a3p1vp",
    Huce71: "fz5stix",
    mc9l5x: "ftgm304",
    sj55zd: "f19n0e5"
  },
  chartTitleLeft4pMargin: {
    jrapky: "fd1gkq"
  },
  chartTitleLeft5pMargin: {
    jrapky: "f475ppk"
  },
  chartTitleRight: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fkhj508",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "f1i3iumi",
    sj55zd: "f19n0e5"
  },
  chartDataTextDenominator: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fkhj508",
    Bhrd7zp: "figsok6",
    Bg96gwp: "f1i3iumi",
    sj55zd: "f19n0e5"
  },
  benchmarkContainer: {
    qhf8xq: "f10pi13n",
    Bqenvij: "f7u49ja",
    B6of3ja: "f1h5g6v3",
    jrapky: "fmxx68s"
  },
  triangle: {
    a9b677: "f3tsq5r",
    Bqenvij: "fniina8",
    ibv6hh: ["fbdn13q", "fdo4c7f"],
    wvpqe5: ["fjik90z", "fcdblym"],
    zhjwy3: ["f1pdflbu", "f11589ue"],
    Bekrc4i: ["fdo4c7f", "fbdn13q"],
    vrafjx: ["fcdblym", "fjik90z"],
    h3c5rm: ["f11589ue", "f1pdflbu"],
    B4j52fo: "f1st6vq5",
    icvyot: "fzkkow9",
    g2u3we: "f1bpgvfk",
    jrapky: "f8l5zjj",
    qhf8xq: "f1euv43f"
  },
  barLabel: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "fwrc4pm",
    Bkfmm31: "fhuob2q",
    Bpvj6i6: "f1l20po4"
  },
  chartWrapper40ppadding: {
    z189sj: ["f1emi2st", "f53ee3v"]
  },
  chartWrapper0ppadding: {
    z189sj: ["flk2ux3", "fkl3uby"]
  },
  legendContainer: {
    z8tnut: "fikn0iw"
  }
}, {
  d: [".f22iagw{display:flex;}", ".f1vx9l62{flex-direction:column;}", ".fly5x3f{width:100%;}", ".f1rfdi54{margin-bottom:var(--spacingVerticalMNudge);}", ".fbxl46w{margin-bottom:var(--spacingVerticalL);}", ".fvblgha{height:12px;}", ".ftgm304{display:block;}", [".f1gl81tg{overflow:visible;}", {
    p: -1
  }], ".fk6fouc{font-family:var(--fontFamilyBase);}", ".fy9rknc{font-size:var(--fontSizeBase200);}", ".figsok6{font-weight:var(--fontWeightRegular);}", ".fwrc4pm{line-height:var(--lineHeightBase200);}", ".f1869bpl{justify-content:space-between;}", ".f1cmbuwj{text-overflow:ellipsis;}", [".f1a3p1vp{overflow:hidden;}", {
    p: -1
  }], ".fz5stix{white-space:nowrap;}", ".f19n0e5{color:var(--colorNeutralForeground1);}", ".fd1gkq{margin-bottom:4px;}", ".f475ppk{margin-bottom:5px;}", ".fkhj508{font-size:var(--fontSizeBase300);}", ".fl43uef{font-weight:var(--fontWeightSemibold);}", ".f1i3iumi{line-height:var(--lineHeightBase300);}", ".f10pi13n{position:relative;}", ".f7u49ja{height:7px;}", ".f1h5g6v3{margin-top:-3px;}", ".fmxx68s{margin-bottom:-1px;}", ".f3tsq5r{width:0;}", ".fniina8{height:0;}", ".fbdn13q{border-left-width:4px;}", ".fdo4c7f{border-right-width:4px;}", ".fjik90z{border-left-style:solid;}", ".fcdblym{border-right-style:solid;}", ".f1pdflbu{border-left-color:transparent;}", ".f11589ue{border-right-color:transparent;}", ".f1st6vq5{border-top-width:7px;}", ".fzkkow9{border-top-style:solid;}", ".f1bpgvfk{border-top-color:var(--colorPaletteBlueBorderActive);}", ".f8l5zjj{margin-bottom:var(--spacingVerticalXS);}", ".f1euv43f{position:absolute;}", ".fhuob2q{fill:var(--colorNeutralForeground1);}", ".f1emi2st{padding-right:40p;}", ".f53ee3v{padding-left:40p;}", ".flk2ux3{padding-right:var(--spacingHorizontalNone);}", ".fkl3uby{padding-left:var(--spacingHorizontalNone);}", ".fikn0iw{padding-top:var(--spacingVerticalL);}"],
  m: [["@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f1l20po4{stroke:CanvasText;}}", {
    m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
  }]]
});
/**
 * Apply styling to the Carousel slots based on the state
 */
export const useHorizontalBarChartStyles = props => {
  var _props_styles, _props_styles1, _props_styles2, _props_styles3, _props_styles4, _props_styles5, _props_styles6, _props_styles7, _props_styles8, _props_styles9, _props_styles10, _props_styles11, _props_styles12;
  const {
    className,
    showTriangle,
    variant,
    hideLabels
  } = props; // ToDo - width, barHeight is non enumerable. Need to be used inline.
  const baseStyles = useStyles();
  return {
    root: mergeClasses(hbcClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
    items: mergeClasses(hbcClassNames.items, showTriangle || variant === HorizontalBarChartVariant.AbsoluteScale ? baseStyles.items16pMargin : baseStyles.items10pMargin, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.items),
    chart: mergeClasses(hbcClassNames.chart, baseStyles.chart, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.chart),
    chartTitle: mergeClasses(hbcClassNames.chartTitle, baseStyles.chartTitle, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.chartTitle),
    barWrapper: mergeClasses(hbcClassNames.barWrapper, baseStyles.barWrapper, (_props_styles4 = props.styles) === null || _props_styles4 === void 0 ? void 0 : _props_styles4.barWrapper),
    chartTitleLeft: mergeClasses(hbcClassNames.chartTitleLeft, baseStyles.chartTitleLeft, variant === HorizontalBarChartVariant.AbsoluteScale ? baseStyles.chartTitleLeft4pMargin : baseStyles.chartTitleLeft5pMargin, (_props_styles5 = props.styles) === null || _props_styles5 === void 0 ? void 0 : _props_styles5.chartTitleLeft),
    chartTitleRight: mergeClasses(hbcClassNames.chartTitleRight, baseStyles.chartTitleRight, (_props_styles6 = props.styles) === null || _props_styles6 === void 0 ? void 0 : _props_styles6.chartTitleRight),
    chartDataTextDenominator: mergeClasses(hbcClassNames.chartDataTextDenominator, baseStyles.chartDataTextDenominator, (_props_styles7 = props.styles) === null || _props_styles7 === void 0 ? void 0 : _props_styles7.chartDataTextDenominator),
    benchmarkContainer: mergeClasses(hbcClassNames.benchmarkContainer, baseStyles.benchmarkContainer, (_props_styles8 = props.styles) === null || _props_styles8 === void 0 ? void 0 : _props_styles8.benchmarkContainer),
    triangle: mergeClasses(hbcClassNames.triangle, baseStyles.triangle, (_props_styles9 = props.styles) === null || _props_styles9 === void 0 ? void 0 : _props_styles9.triangle),
    barLabel: mergeClasses(hbcClassNames.barLabel, baseStyles.barLabel, (_props_styles10 = props.styles) === null || _props_styles10 === void 0 ? void 0 : _props_styles10.barLabel),
    chartWrapper: mergeClasses(hbcClassNames.chartWrapper, variant === HorizontalBarChartVariant.AbsoluteScale && !hideLabels ? baseStyles.chartWrapper40ppadding : baseStyles.chartWrapper0ppadding, (_props_styles11 = props.styles) === null || _props_styles11 === void 0 ? void 0 : _props_styles11.chartWrapper),
    legendContainer: mergeClasses(hbcClassNames.legendContainer, baseStyles.legendContainer, (_props_styles12 = props.styles) === null || _props_styles12 === void 0 ? void 0 : _props_styles12.legendContainer)
  };
};
