// @fluentui/react-charts/lib/components/DonutChart/Pie/usePieStyles.styles.js
'use client';

import { tokens, typographyStyles } from '@fluentui/react-theme';
import { __styles, mergeClasses } from '@griffel/react';
/**
 * @internal
 */
export const donutPieClassNames = {
  root: 'fui-donut-pie__root',
  insideDonutString: 'fui-donut-pie__insideDonutString'
};
/**
 * Base Styles
 */
const useStyles = /*#__PURE__*/__styles({
  root: {},
  insideDonutString: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fojgt09",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "fcen8rp",
    Bkfmm31: "fhuob2q",
    Bvjb7m6: "fhv2zbx"
  }
}, {
  d: [".fk6fouc{font-family:var(--fontFamilyBase);}", ".fojgt09{font-size:var(--fontSizeHero700);}", ".fl43uef{font-weight:var(--fontWeightSemibold);}", ".fcen8rp{line-height:var(--lineHeightHero700);}", ".fhuob2q{fill:var(--colorNeutralForeground1);}", ".fhv2zbx{forced-color-adjust:auto;}"]
});
/**
 * Apply styling to the Pie inside donut chart component
 */
export const usePieStyles = props => {
  var _props_styles, _props_styles1;
  const {
    className
  } = props;
  const baseStyles = useStyles();
  return {
    root: mergeClasses(donutPieClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
    insideDonutString: mergeClasses(donutPieClassNames.insideDonutString, baseStyles.insideDonutString, className, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.insideDonutString)
  };
};
