// @fluentui/react-charts/lib-commonjs/components/DonutChart/useDonutChartStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    donutClassNames: function() {
        return donutClassNames;
    },
    useDonutChartStyles: function() {
        return useDonutChartStyles;
    }
});
const _react = require("@griffel/react");
const donutClassNames = {
    root: 'fui-donut__root',
    chart: 'fui-donut__chart',
    legendContainer: 'fui-donut__legendContainer',
    chartWrapper: 'fui-donut__chartWrapper',
    axisAnnotation: 'fui-donut__axisAnnotation'
};
/**
 * Base Styles
 */ const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508",
        Bhrd7zp: "figsok6",
        Bg96gwp: "f1i3iumi",
        mc9l5x: "f22iagw",
        Beiy3e4: "f1vx9l62",
        a9b677: "fly5x3f",
        Bqenvij: "f1l02sjl"
    },
    chart: {
        B7ck84d: "f1e4lqlz",
        ek0pe7: "fzm18in",
        mc9l5x: "ftgm304",
        B68tc82: 0,
        Bmxbyg5: 0,
        Bpg54ce: "f1gl81tg"
    },
    legendContainer: {
        z8tnut: "fikn0iw",
        a9b677: "fly5x3f"
    },
    axisAnnotation: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "f13mqy1h",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fcpl73t",
        B80ckks: "fmd4ok8",
        fsow6f: "f17mccla",
        sj55zd: "fkfq4zb",
        Bkfmm31: "fhuob2q",
        a6j6cd: "folz2wl"
    }
}, {
    d: [
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fkhj508{font-size:var(--fontSizeBase300);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".f1i3iumi{line-height:var(--lineHeightBase300);}",
        ".f22iagw{display:flex;}",
        ".f1vx9l62{flex-direction:column;}",
        ".fly5x3f{width:100%;}",
        ".f1l02sjl{height:100%;}",
        ".f1e4lqlz{box-sizing:content-box;}",
        ".fzm18in{alignment-adjust:center;}",
        ".ftgm304{display:block;}",
        [
            ".f1gl81tg{overflow:visible;}",
            {
                p: -1
            }
        ],
        ".fikn0iw{padding-top:var(--spacingVerticalL);}",
        ".f13mqy1h{font-size:var(--fontSizeBase100);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fcpl73t{line-height:var(--lineHeightBase100);}",
        ".fmd4ok8{font-style:normal;}",
        ".f17mccla{text-align:center;}",
        ".fkfq4zb{color:var(--colorNeutralForeground2);}",
        ".fhuob2q{fill:var(--colorNeutralForeground1);}"
    ],
    m: [
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.folz2wl{fill:CanvasText;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ]
    ]
});
const useDonutChartStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2, _props_styles3, _props_styles4;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(donutClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        chart: (0, _react.mergeClasses)(donutClassNames.chart, baseStyles.chart, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.chart),
        legendContainer: (0, _react.mergeClasses)(donutClassNames.legendContainer, baseStyles.legendContainer, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.legendContainer),
        chartWrapper: (0, _react.mergeClasses)(donutClassNames.chartWrapper, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.chartWrapper),
        axisAnnotation: (0, _react.mergeClasses)(donutClassNames.axisAnnotation, baseStyles.axisAnnotation, (_props_styles4 = props.styles) === null || _props_styles4 === void 0 ? void 0 : _props_styles4.axisAnnotation)
    };
};
