// @fluentui/react-charts/lib-commonjs/components/AreaChart/useAreaChartStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    areaChartClassNames: function() {
        return areaChartClassNames;
    },
    useAreaChartStyles: function() {
        return useAreaChartStyles;
    }
});
const _react = require("@griffel/react");
const _index = require("../../utilities/index");
const areaChartClassNames = {
    tooltip: 'fui-ac__tooltip',
    root: '',
    xAxis: '',
    yAxis: '',
    legendContainer: '',
    hover: '',
    descriptionMessage: '',
    axisTitle: '',
    chartTitle: '',
    opacityChangeOnHover: '',
    shapeStyles: '',
    chartWrapper: '',
    svgTooltip: '',
    chart: '',
    axisAnnotation: ''
};
const useStyles = (0, _react.makeStyles)({
    tooltip: (0, _index.getTooltipStyle)()
});
const useAreaChartStyles = (props)=>{
    const baseStyles = useStyles();
    return {
        tooltip: (0, _react.mergeClasses)(areaChartClassNames.tooltip, baseStyles.tooltip /*props.styles?.tooltip*/ )
    };
};
