// @fluentui/react-charts/lib-commonjs/components/AreaChart/useAreaChartStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    areaChartClassNames: function() {
        return areaChartClassNames;
    },
    useAreaChartStyles: function() {
        return useAreaChartStyles;
    }
});
const _react = require("@griffel/react");
const areaChartClassNames = {
    tooltip: 'fui-ac__tooltip',
    root: '',
    xAxis: '',
    yAxis: '',
    legendContainer: '',
    hover: '',
    descriptionMessage: '',
    axisTitle: '',
    chartTitle: '',
    opacityChangeOnHover: '',
    shapeStyles: '',
    chartWrapper: '',
    svgTooltip: '',
    chart: '',
    axisAnnotation: ''
};
const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    tooltip: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508",
        Bhrd7zp: "figsok6",
        Bg96gwp: "f1i3iumi",
        mc9l5x: "f22iagw",
        Beiy3e4: "f1vx9l62",
        z8tnut: "f17mpqex",
        z189sj: [
            "f1vdfbxk",
            "f1f5gg8d"
        ],
        Byoj8tv: "fdvome7",
        uwmqm3: [
            "f1f5gg8d",
            "f1vdfbxk"
        ],
        qhf8xq: "f1euv43f",
        fsow6f: "f17mccla",
        Bhzewxz: "fr6rvge",
        De3pzq: "fxugw4r",
        Beyfa6y: 0,
        Bbmb7ep: 0,
        Btl43ni: 0,
        B7oj6ja: 0,
        Dimara: "fq9zq91",
        Bkecrkj: "f1aehjj5",
        sj55zd: "f19n0e5"
    }
}, {
    d: [
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fkhj508{font-size:var(--fontSizeBase300);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".f1i3iumi{line-height:var(--lineHeightBase300);}",
        ".f22iagw{display:flex;}",
        ".f1vx9l62{flex-direction:column;}",
        ".f17mpqex{padding-top:var(--spacingHorizontalS);}",
        ".f1vdfbxk{padding-right:var(--spacingHorizontalS);}",
        ".f1f5gg8d{padding-left:var(--spacingHorizontalS);}",
        ".fdvome7{padding-bottom:var(--spacingHorizontalS);}",
        ".f1euv43f{position:absolute;}",
        ".f17mccla{text-align:center;}",
        ".fr6rvge{top:var(--spacingVerticalNone);}",
        ".fxugw4r{background-color:var(--colorNeutralBackground1);}",
        [
            ".fq9zq91{border-radius:var(--borderRadiusSmall);}",
            {
                p: -1
            }
        ],
        ".f1aehjj5{pointer-events:none;}",
        ".f19n0e5{color:var(--colorNeutralForeground1);}"
    ]
});
const useAreaChartStyles = (props)=>{
    const baseStyles = useStyles();
    return {
        tooltip: (0, _react.mergeClasses)(areaChartClassNames.tooltip, baseStyles.tooltip /*props.styles?.tooltip*/ )
    };
};
