// @fluentui/react-charts/lib-commonjs/components/DonutChart/Arc/useArcStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    donutArcClassNames: function() {
        return donutArcClassNames;
    },
    useArcStyles: function() {
        return useArcStyles;
    }
});
const _react = require("@griffel/react");
const donutArcClassNames = {
    root: 'fui-donut-arc__root',
    focusRing: 'fui-donut-arc__focusRing',
    arcLabel: 'fui-donut-arc__arcLabel'
};
/**
 * Base Styles
 */ const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        Bceei9c: "f158kwzp",
        Bw0xxkn: 0,
        oeaueh: 0,
        Bpd4iqm: 0,
        Befb4lg: "f1391ufr",
        ojy3ng: "f1yuyku4",
        B4ncu3z: "f138xzbb",
        zxy5zd: [
            "fny96iz",
            "fi0bkha"
        ],
        bkw2yb: "f1jv8gm3",
        Bjksbd0: [
            "fi0bkha",
            "fny96iz"
        ]
    },
    focusRing: {
        ojy3ng: "frcfrt5",
        Be5yapy: "f9nnx8l",
        Bkfmm31: "f1au8mb3"
    },
    arcLabel: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fwrc4pm",
        Bkfmm31: "fhuob2q",
        Bpvj6i6: "f1l20po4"
    }
}, {
    d: [
        ".f158kwzp{cursor:default;}",
        [
            ".f1391ufr{outline:transparent;}",
            {
                p: -1
            }
        ],
        ".f1yuyku4{stroke:var(--colorNeutralBackground1);}",
        ".f138xzbb selectors::-moz-focus-inner{border-top-width:0;}",
        ".fny96iz selectors::-moz-focus-inner{border-right-width:0;}",
        ".fi0bkha selectors::-moz-focus-inner{border-left-width:0;}",
        ".f1jv8gm3 selectors::-moz-focus-inner{border-bottom-width:0;}",
        ".frcfrt5{stroke:var(--colorStrokeFocus2);}",
        ".f9nnx8l{stroke-width:var(--strokeWidthThickest);}",
        ".f1au8mb3{fill:transparent;}",
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fy9rknc{font-size:var(--fontSizeBase200);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fwrc4pm{line-height:var(--lineHeightBase200);}",
        ".fhuob2q{fill:var(--colorNeutralForeground1);}"
    ],
    m: [
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f1l20po4{stroke:CanvasText;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ]
    ]
});
const useArcStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2;
    const { className } = props;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(donutArcClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        focusRing: (0, _react.mergeClasses)(donutArcClassNames.focusRing, baseStyles.focusRing, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.focusRing),
        arcLabel: (0, _react.mergeClasses)(donutArcClassNames.arcLabel, baseStyles.arcLabel, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.arcLabel)
    };
};
