// @fluentui/react-charts/lib-commonjs/components/VerticalBarChart/useVerticalBarChartStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    useVerticalBarChartStyles: function() {
        return useVerticalBarChartStyles;
    },
    verticalbarchartClassNames: function() {
        return verticalbarchartClassNames;
    }
});
const _react = require("@griffel/react");
const verticalbarchartClassNames = {
    opacityChangeOnHover: 'fui-vbc__opacityChangeOnHover',
    tooltip: 'fui-vbc__tooltip',
    barLabel: 'fui-vbc__barLabel',
    lineBorder: 'fui-vbc_lineBorder',
    root: '',
    xAxis: '',
    yAxis: '',
    legendContainer: '',
    hover: '',
    descriptionMessage: '',
    axisTitle: '',
    chartTitle: '',
    shapeStyles: '',
    chartWrapper: '',
    svgTooltip: '',
    chart: '',
    axisAnnotation: ''
};
const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    opacityChangeOnHover: {},
    tooltip: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508",
        Bhrd7zp: "figsok6",
        Bg96gwp: "f1i3iumi",
        mc9l5x: "f22iagw",
        Beiy3e4: "f1vx9l62",
        z8tnut: "f17mpqex",
        z189sj: [
            "f1vdfbxk",
            "f1f5gg8d"
        ],
        Byoj8tv: "fdvome7",
        uwmqm3: [
            "f1f5gg8d",
            "f1vdfbxk"
        ],
        qhf8xq: "f1euv43f",
        fsow6f: "f17mccla",
        Bhzewxz: "fr6rvge",
        De3pzq: "fxugw4r",
        Beyfa6y: 0,
        Bbmb7ep: 0,
        Btl43ni: 0,
        B7oj6ja: 0,
        Dimara: "fq9zq91",
        Bkecrkj: "f1aehjj5",
        sj55zd: "f19n0e5"
    },
    barLabel: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fwrc4pm",
        Bkfmm31: "fhuob2q",
        Bvjb7m6: "fhv2zbx"
    },
    lineBorder: {
        ojy3ng: "f1yuyku4",
        Bpvj6i6: "fdvt8h6"
    }
}, {
    d: [
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fkhj508{font-size:var(--fontSizeBase300);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".f1i3iumi{line-height:var(--lineHeightBase300);}",
        ".f22iagw{display:flex;}",
        ".f1vx9l62{flex-direction:column;}",
        ".f17mpqex{padding-top:var(--spacingHorizontalS);}",
        ".f1vdfbxk{padding-right:var(--spacingHorizontalS);}",
        ".f1f5gg8d{padding-left:var(--spacingHorizontalS);}",
        ".fdvome7{padding-bottom:var(--spacingHorizontalS);}",
        ".f1euv43f{position:absolute;}",
        ".f17mccla{text-align:center;}",
        ".fr6rvge{top:var(--spacingVerticalNone);}",
        ".fxugw4r{background-color:var(--colorNeutralBackground1);}",
        [
            ".fq9zq91{border-radius:var(--borderRadiusSmall);}",
            {
                p: -1
            }
        ],
        ".f1aehjj5{pointer-events:none;}",
        ".f19n0e5{color:var(--colorNeutralForeground1);}",
        ".fy9rknc{font-size:var(--fontSizeBase200);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fwrc4pm{line-height:var(--lineHeightBase200);}",
        ".fhuob2q{fill:var(--colorNeutralForeground1);}",
        ".fhv2zbx{forced-color-adjust:auto;}",
        ".f1yuyku4{stroke:var(--colorNeutralBackground1);}"
    ],
    m: [
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.fdvt8h6{stroke:Canvas;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ]
    ]
});
const useVerticalBarChartStyles = (props)=>{
    const baseStyles = useStyles();
    return {
        opacityChangeOnHover: (0, _react.mergeClasses)(verticalbarchartClassNames.opacityChangeOnHover, baseStyles.opacityChangeOnHover /*props.styles?.opacityChangeOnHover*/ ),
        tooltip: (0, _react.mergeClasses)(verticalbarchartClassNames.tooltip, baseStyles.tooltip /*props.styles?.tooltip*/ ),
        barLabel: (0, _react.mergeClasses)(verticalbarchartClassNames.barLabel, baseStyles.barLabel /*props.styles?.barLabel*/ ),
        lineBorder: (0, _react.mergeClasses)(verticalbarchartClassNames.lineBorder, baseStyles.lineBorder /*props.styles?.lineBorder*/ )
    };
};
