// @fluentui/react-charts/lib/components/LineChart/eventAnnotation/LabelLink.js
'use client';
import * as React from 'react';
import { Textbox } from './Textbox';
import { getColorFromToken } from '../../../utilities/colors';
import { tokens } from '@fluentui/react-theme';
export const LabelLink = (props)=>{
    const gRef = React.useRef(null);
    const [showCard, setShowCard] = React.useState(false);
    const onClick = ()=>setShowCard(true);
    let callout = null;
    if (showCard) {
        const cards = props.labelDef.aggregatedIdx.map((i)=>props.lineDefs[i].onRenderCard).filter((c)=>!!c);
        if (cards.length > 0) {
            callout = null;
        // TODO - need to replace callout with popover
        /*callout = {
        /* <Callout
          target={gRef.current}
          // eslint-disable-next-line react/jsx-no-bind
          onDismiss={onDismiss}
          setInitialFocus={true}
          role="dialog"
        >
          <FocusZone isCircularNavigation={true} direction={FocusZoneDirection.vertical}>
            <List<() => React.ReactNode>
              items={cards}
              // eslint-disable-next-line react/jsx-no-bind
              onRenderCell={onRenderCell}
            />
          </FocusZone>
        </Callout>
      };*/ }
    }
    let text;
    const fill = props.textColor ? getColorFromToken(props.textColor, false) : tokens.colorNeutralForeground1;
    if (props.labelDef.aggregatedIdx.length === 1) {
        text = props.lineDefs[props.labelDef.aggregatedIdx[0]].event;
    } else {
        text = props.mergedLabel(props.labelDef.aggregatedIdx.length);
    }
    return /*#__PURE__*/ React.createElement(React.Fragment, null, /*#__PURE__*/ React.createElement("g", {
        ref: gRef,
        onClick: onClick,
        "data-is-focusable": false,
        style: {
            cursor: 'pointer'
        },
        role: "button"
    }, /*#__PURE__*/ React.createElement(Textbox, {
        text: text,
        x: props.labelDef.x,
        y: props.textY,
        width: props.textWidth,
        lineHeight: props.textLineHeight,
        textAnchor: props.labelDef.anchor,
        fontSize: props.textFontSize,
        fill: fill
    })), callout);
};
