// @fluentui/react-charts/lib-commonjs/components/GroupedVerticalBarChart/useGroupedVerticalBarChartStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    groupedVerticalBarChartClassNames: function() {
        return groupedVerticalBarChartClassNames;
    },
    useGroupedVerticalBarChartStyles_unstable: function() {
        return useGroupedVerticalBarChartStyles_unstable;
    }
});
const _react = require("@griffel/react");
const _index = require("../../utilities/index");
const groupedVerticalBarChartClassNames = {
    opacityChangeOnHover: 'fui-gvbc**opacityChangeOnHover',
    tooltip: 'fui-gvbc**tooltip',
    barLabel: 'fui-gvbc**barLabel',
    root: '',
    xAxis: '',
    yAxis: '',
    legendContainer: '',
    hover: '',
    descriptionMessage: '',
    axisTitle: '',
    chartTitle: '',
    shapeStyles: '',
    chartWrapper: '',
    svgTooltip: '',
    chart: '',
    axisAnnotation: ''
};
const useStyles = (0, _react.makeStyles)({
    opacityChangeOnHover: {
        cursor: 'default'
    },
    tooltip: (0, _index.getTooltipStyle)(),
    barLabel: (0, _index.getBarLabelStyle)()
});
const useGroupedVerticalBarChartStyles_unstable = (props)=>{
    const baseStyles = useStyles();
    return {
        opacityChangeOnHover: (0, _react.mergeClasses)(groupedVerticalBarChartClassNames.opacityChangeOnHover, baseStyles.opacityChangeOnHover /*props.styles?.opacityChangeOnHover*/ ),
        tooltip: (0, _react.mergeClasses)(groupedVerticalBarChartClassNames.tooltip, baseStyles.tooltip /*props.styles?.tooltip*/ ),
        barLabel: (0, _react.mergeClasses)(groupedVerticalBarChartClassNames.barLabel, baseStyles.barLabel /*props.styles?.barLabel*/ )
    };
};
