// @fluentui/react-charts/lib-commonjs/components/GaugeChart/useGaugeChartStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    gaugeChartClassNames: function() {
        return gaugeChartClassNames;
    },
    useGaugeChartStyles: function() {
        return useGaugeChartStyles;
    }
});
const _reacttheme = require("@fluentui/react-theme");
const _react = require("@griffel/react");
const gaugeChartClassNames = {
    root: 'fui-gc__root',
    chart: 'fui-gc__chart',
    limits: 'fui-gc__limits',
    chartValue: 'fui-gc__chartValue',
    sublabel: 'fui-gc__sublabel',
    needle: 'fui-gc__needle',
    chartTitle: 'fui-gc__chartTitle',
    segment: 'fui-gc__segment',
    gradientSegment: 'fui-gc__gradientSegment',
    calloutContentRoot: 'fui-gc__calloutContentRoot',
    calloutDateTimeContainer: 'fui-gc__calloutDateTimeContainer',
    calloutContentX: 'fui-gc__calloutContentX',
    calloutBlockContainer: 'fui-gc__calloutBlockContainer',
    shapeStyles: 'fui-gc__shapeStyles',
    calloutlegendText: 'fui-gc__calloutlegendText',
    calloutContentY: 'fui-gc__calloutContentY',
    descriptionMessage: 'fui-gc__descriptionMessage',
    calloutInfoContainer: '',
    legendsContainer: 'fui-gc__legendsContainer',
    chartWrapper: 'fui-gc__chartWrapper'
};
const useStyles = (0, _react.makeStyles)({
    root: {
        ..._reacttheme.typographyStyles.body1,
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        width: '100%',
        height: '100%'
    },
    chart: {
        display: 'block'
    },
    limits: {
        ..._reacttheme.typographyStyles.caption1Strong,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    },
    chartValue: {
        fontWeight: _reacttheme.tokens.fontWeightSemibold,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    },
    sublabel: {
        ..._reacttheme.typographyStyles.caption1Strong,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    },
    needle: {
        fill: _reacttheme.tokens.colorNeutralForeground1,
        stroke: _reacttheme.tokens.colorNeutralBackground1
    },
    chartTitle: {
        ..._reacttheme.typographyStyles.caption1,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    },
    segment: {
        outline: 'none',
        stroke: _reacttheme.tokens.colorNeutralStroke1
    },
    gradientSegment: {
        width: '100%',
        height: '100%'
    },
    calloutContentRoot: {
        display: 'grid',
        overflow: 'hidden',
        backgroundColor: _reacttheme.tokens.colorNeutralBackground1,
        backgroundBlendMode: 'normal, luminosity'
    },
    calloutDateTimeContainer: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    calloutContentX: {
        ..._reacttheme.typographyStyles.caption1,
        lineHeight: '16px',
        opacity: '0.85'
    },
    calloutBlockContainer: {
        ..._reacttheme.typographyStyles.body1Strong,
        marginTop: '13px',
        color: _reacttheme.tokens.colorNeutralForeground1,
        paddingLeft: '8px',
        display: 'block',
        forcedColorAdjust: 'none'
    },
    shapeStyles: {
        marginRight: '8px'
    },
    calloutlegendText: {
        ..._reacttheme.typographyStyles.caption1,
        lineHeight: '16px',
        color: _reacttheme.tokens.colorNeutralForeground2,
        forcedColorAdjust: 'auto'
    },
    calloutContentY: {
        ..._reacttheme.typographyStyles.body1Strong,
        lineHeight: '22px',
        forcedColorAdjust: 'auto'
    },
    descriptionMessage: {
        ..._reacttheme.typographyStyles.caption1,
        color: _reacttheme.tokens.colorNeutralForeground1,
        marginTop: '10px',
        paddingTop: '10px',
        borderTop: `1px solid ${_reacttheme.tokens.colorNeutralStroke1}`
    },
    legendsContainer: {
        width: '100%'
    }
});
const useGaugeChartStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2, _props_styles3, _props_styles4, _props_styles5, _props_styles6, _props_styles7, _props_styles8, _props_styles9, _props_styles10, _props_styles11, _props_styles12, _props_styles13, _props_styles14, _props_styles15, _props_styles16, _props_styles17, _props_styles18;
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(gaugeChartClassNames.root, baseStyles.root, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        chart: (0, _react.mergeClasses)(gaugeChartClassNames.chart, baseStyles.chart, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.chart),
        limits: (0, _react.mergeClasses)(gaugeChartClassNames.limits, baseStyles.limits, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.limits),
        chartValue: (0, _react.mergeClasses)(gaugeChartClassNames.chartValue, baseStyles.chartValue, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.chartValue),
        sublabel: (0, _react.mergeClasses)(gaugeChartClassNames.sublabel, baseStyles.sublabel, (_props_styles4 = props.styles) === null || _props_styles4 === void 0 ? void 0 : _props_styles4.sublabel),
        needle: (0, _react.mergeClasses)(gaugeChartClassNames.needle, baseStyles.needle, (_props_styles5 = props.styles) === null || _props_styles5 === void 0 ? void 0 : _props_styles5.needle),
        chartTitle: (0, _react.mergeClasses)(gaugeChartClassNames.chartTitle, baseStyles.chartTitle, (_props_styles6 = props.styles) === null || _props_styles6 === void 0 ? void 0 : _props_styles6.chartTitle),
        segment: (0, _react.mergeClasses)(gaugeChartClassNames.segment, baseStyles.segment, (_props_styles7 = props.styles) === null || _props_styles7 === void 0 ? void 0 : _props_styles7.segment),
        gradientSegment: (0, _react.mergeClasses)(gaugeChartClassNames.gradientSegment, baseStyles.gradientSegment, (_props_styles8 = props.styles) === null || _props_styles8 === void 0 ? void 0 : _props_styles8.gradientSegment),
        calloutContentRoot: (0, _react.mergeClasses)(gaugeChartClassNames.calloutContentRoot, baseStyles.calloutContentRoot, (_props_styles9 = props.styles) === null || _props_styles9 === void 0 ? void 0 : _props_styles9.calloutContentRoot),
        calloutDateTimeContainer: (0, _react.mergeClasses)(gaugeChartClassNames.calloutDateTimeContainer, baseStyles.calloutDateTimeContainer, (_props_styles10 = props.styles) === null || _props_styles10 === void 0 ? void 0 : _props_styles10.calloutDateTimeContainer),
        calloutContentX: (0, _react.mergeClasses)(gaugeChartClassNames.calloutContentX, baseStyles.calloutContentX, (_props_styles11 = props.styles) === null || _props_styles11 === void 0 ? void 0 : _props_styles11.calloutContentX),
        calloutBlockContainer: (0, _react.mergeClasses)(gaugeChartClassNames.calloutBlockContainer, baseStyles.calloutBlockContainer, (_props_styles12 = props.styles) === null || _props_styles12 === void 0 ? void 0 : _props_styles12.calloutBlockContainer),
        shapeStyles: (0, _react.mergeClasses)(gaugeChartClassNames.shapeStyles, baseStyles.shapeStyles, (_props_styles13 = props.styles) === null || _props_styles13 === void 0 ? void 0 : _props_styles13.shapeStyles),
        calloutlegendText: (0, _react.mergeClasses)(gaugeChartClassNames.calloutlegendText, baseStyles.calloutlegendText, (_props_styles14 = props.styles) === null || _props_styles14 === void 0 ? void 0 : _props_styles14.calloutlegendText),
        calloutContentY: (0, _react.mergeClasses)(gaugeChartClassNames.calloutContentY, baseStyles.calloutContentY, (_props_styles15 = props.styles) === null || _props_styles15 === void 0 ? void 0 : _props_styles15.calloutContentY),
        descriptionMessage: (0, _react.mergeClasses)(gaugeChartClassNames.descriptionMessage, baseStyles.descriptionMessage, (_props_styles16 = props.styles) === null || _props_styles16 === void 0 ? void 0 : _props_styles16.descriptionMessage),
        chartWrapper: (0, _react.mergeClasses)(gaugeChartClassNames.chartWrapper, (_props_styles17 = props.styles) === null || _props_styles17 === void 0 ? void 0 : _props_styles17.chartWrapper),
        legendsContainer: (0, _react.mergeClasses)(gaugeChartClassNames.legendsContainer, baseStyles.legendsContainer, (_props_styles18 = props.styles) === null || _props_styles18 === void 0 ? void 0 : _props_styles18.legendsContainer)
    };
};
