// @fluentui/react-charts/lib-commonjs/components/Sparkline/useSparklineStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    sparklineClassNames: function() {
        return sparklineClassNames;
    },
    useSparklineStyles: function() {
        return useSparklineStyles;
    }
});
const _react = require("@griffel/react");
const _reacttheme = require("@fluentui/react-theme");
const sparklineClassNames = {
    inlineBlock: 'fui-sprk__inlineBlock',
    valueText: 'fui-sprk__valueText'
};
/**
 * Base Styles
 */ const useStyles = (0, _react.makeStyles)({
    inlineBlock: {
        display: 'inline'
    },
    valueText: {
        ..._reacttheme.typographyStyles.caption1,
        fill: _reacttheme.tokens.colorNeutralForeground1,
        forcedColorAdjust: 'auto'
    }
});
const useSparklineStyles = (props)=>{
    var _props_styles, _props_styles1;
    const baseStyles = useStyles();
    return {
        inlineBlock: (0, _react.mergeClasses)(sparklineClassNames.inlineBlock, baseStyles.inlineBlock, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.inlineBlock),
        valueText: (0, _react.mergeClasses)(sparklineClassNames.valueText, baseStyles.valueText, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.valueText)
    };
};
