// @fluentui/react-charts/lib-commonjs/components/Legends/shape.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Shape", {
    enumerable: true,
    get: function() {
        return Shape;
    }
});
const _interop_require_wildcard = require("@swc/helpers/_/_interop_require_wildcard");
const _react = /*#__PURE__*/ _interop_require_wildcard._(require("react"));
const _utilities = require("../../utilities/utilities");
const pointPath = {
    [`${_utilities.Points[_utilities.Points.circle]}`]: 'M1 6 A5 5 0 1 0  12 6 M1 6 A5 5 0 0 1  12 6',
    [`${_utilities.Points[_utilities.Points.square]}`]: 'M1 1 L12 1 L12 12  L1 12 L1 1 Z',
    [`${_utilities.Points[_utilities.Points.triangle]}`]: 'M6 10L8.74228e-07 -1.04907e-06L12 0L6 10Z',
    [`${_utilities.Points[_utilities.Points.pyramid]}`]: 'M6 10L8.74228e-07 -1.04907e-06L12 0L6 10Z',
    [`${_utilities.Points[_utilities.Points.diamond]}`]: 'M2 2 L10 2 L10 10  L2 10 L2 2 Z',
    [`${_utilities.Points[_utilities.Points.hexagon]}`]: 'M9 0H3L0 5L3 10H9L12 5L9 0Z',
    [`${_utilities.Points[_utilities.Points.pentagon]}`]: 'M6.06061 0L0 4.21277L2.30303 11H9.69697L12 4.21277L6.06061 0Z',
    [`${_utilities.Points[_utilities.Points.octagon]}`]: 'M7.08333 0H2.91667L0 2.91667V7.08333L2.91667 10H7.08333L10 7.08333V2.91667L7.08333 0Z',
    [`${_utilities.CustomPoints[_utilities.CustomPoints.dottedLine]}`]: 'M0 6 H3 M5 6 H8 M10 6 H13'
};
const Shape = /*#__PURE__*/ _react.forwardRef(({ svgProps, pathProps, shape, classNameForNonSvg, style }, forwardedRef)=>{
    if (Object.keys(pointPath).indexOf(shape) === -1) {
        return /*#__PURE__*/ _react.createElement("div", {
            className: classNameForNonSvg,
            style: style
        });
    }
    return /*#__PURE__*/ _react.createElement("svg", {
        width: 14,
        height: 14,
        viewBox: '-1 -1 14 14',
        ...(0, _utilities.getSecureProps)(svgProps),
        transform: `rotate(${shape === _utilities.Points[_utilities.Points.diamond] ? 45 : shape === _utilities.Points[_utilities.Points.pyramid] ? 180 : 0}, 0, 0)`,
        style: {
            width: '14px',
            height: '14px'
        }
    }, /*#__PURE__*/ _react.createElement("path", {
        d: pointPath[shape],
        ...(0, _utilities.getSecureProps)(pathProps)
    }));
});
