// @fluentui/react-charts/lib-commonjs/components/Legends/useLegendsStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    INACTIVE_LEGEND_TEXT_OPACITY: function() {
        return INACTIVE_LEGEND_TEXT_OPACITY;
    },
    LEGEND_CONTAINER_MARGIN_START: function() {
        return LEGEND_CONTAINER_MARGIN_START;
    },
    LEGEND_CONTAINER_MARGIN_TOP: function() {
        return LEGEND_CONTAINER_MARGIN_TOP;
    },
    LEGEND_HEIGHT: function() {
        return LEGEND_HEIGHT;
    },
    LEGEND_PADDING: function() {
        return LEGEND_PADDING;
    },
    LEGEND_SHAPE_BORDER: function() {
        return LEGEND_SHAPE_BORDER;
    },
    LEGEND_SHAPE_MARGIN_END: function() {
        return LEGEND_SHAPE_MARGIN_END;
    },
    LEGEND_SHAPE_SIZE: function() {
        return LEGEND_SHAPE_SIZE;
    },
    legendClassNames: function() {
        return legendClassNames;
    },
    useLegendStyles: function() {
        return useLegendStyles;
    }
});
const _react = require("@griffel/react");
const LEGEND_CONTAINER_MARGIN_TOP = 8;
const LEGEND_CONTAINER_MARGIN_START = 12;
const LEGEND_PADDING = 8;
const LEGEND_HEIGHT = 32;
const LEGEND_SHAPE_BORDER = 1;
const LEGEND_SHAPE_SIZE_WITHOUT_BORDER = 12;
const LEGEND_SHAPE_SIZE = LEGEND_SHAPE_SIZE_WITHOUT_BORDER + LEGEND_SHAPE_BORDER;
const LEGEND_SHAPE_MARGIN_END = 8;
const INACTIVE_LEGEND_TEXT_OPACITY = 0.67;
const legendClassNames = {
    root: 'fui-legend__root',
    legend: 'fui-legend__legend',
    rect: 'fui-legend__rect',
    shape: 'fui-legend__shape',
    triangle: 'fui-legend__triangle',
    text: 'fui-legend__text',
    hoverChange: 'fui-legend__hoverChange',
    resizableArea: 'fui-legend__resizableArea',
    legendContainer: 'fui-legend__legendContainer',
    annotation: 'fui-legend__annotation'
};
const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        Huce71: "fz5stix",
        a9b677: "fly5x3f",
        Bt984gj: "f122n59",
        B6of3ja: "fpzxe1g",
        t21cq0: [
            "f705vk6",
            "fqom4eq"
        ],
        jrapky: "fzpzak3",
        Frg6f3: [
            "fqom4eq",
            "f705vk6"
        ]
    },
    legend: {
        Bt984gj: "f122n59",
        Brf1p80: [
            "fc7pyfw",
            "fgbriwd"
        ],
        Bceei9c: "f1k6fduh",
        icvyot: "f1ern45e",
        vrafjx: [
            "f1n71otn",
            "f1deefiw"
        ],
        oivjwe: "f1h8hb77",
        wvpqe5: [
            "f1deefiw",
            "f1n71otn"
        ],
        Byoj8tv: 0,
        uwmqm3: 0,
        z189sj: 0,
        z8tnut: 0,
        B0ocmuz: "f1t756mo",
        B9bfxx9: "fiwaqkr",
        Bf4jedk: "fy77jfu",
        Bs92zhm: "fuhkvgp",
        B93v9kj: "f15mrrko",
        Ck9m1r: "f1vu5irt",
        mkjk8h: "f1sdboh2"
    },
    rect: {
        zpeayv: "f81dvq6",
        Bcwlslt: "f1a000x3",
        a9b677: "frx94fk",
        t21cq0: [
            "f1phki43",
            "ff9s3yw"
        ],
        Bgfg5da: 0,
        B9xav0g: 0,
        oivjwe: 0,
        Bn0qgzm: 0,
        B4g9neb: 0,
        zhjwy3: 0,
        wvpqe5: 0,
        ibv6hh: 0,
        u1mtju: 0,
        h3c5rm: 0,
        vrafjx: 0,
        Bekrc4i: 0,
        i8vvqc: 0,
        g2u3we: 0,
        icvyot: 0,
        B4j52fo: 0,
        irswps: "fpyyuzi"
    },
    shape: {
        t21cq0: [
            "f1phki43",
            "ff9s3yw"
        ]
    },
    triangle: {
        a9b677: "f3tsq5r",
        Bqenvij: "fniina8",
        ibv6hh: [
            "fk1dseo",
            "f1r9b9o"
        ],
        Bekrc4i: [
            "f1r9b9o",
            "fk1dseo"
        ],
        B4j52fo: "f1cm5upi",
        t21cq0: [
            "f1vcna3q",
            "foyynoy"
        ]
    },
    text: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fwrc4pm",
        sj55zd: "f19n0e5",
        Bvjb7m6: "fhv2zbx"
    },
    hoverChange: {
        a9b677: "frx94fk",
        Bqenvij: "fvblgha",
        t21cq0: [
            "f1vcna3q",
            "foyynoy"
        ],
        B4j52fo: "f1d7653x",
        Bekrc4i: [
            "f1maxq7k",
            "f16g4l54"
        ],
        Bn0qgzm: "f1rzi0e7",
        ibv6hh: [
            "f16g4l54",
            "f1maxq7k"
        ]
    },
    resizableArea: {
        qhf8xq: "f10pi13n",
        fsow6f: [
            "f1o700av",
            "fes3tcz"
        ],
        Bz10aip: [
            "f15ivbov",
            "fjwp9lj"
        ],
        Bhzewxz: "f198g47y",
        oyh7mz: [
            "f1ktbn1t",
            "ffenbu1"
        ],
        Bf4jedk: "f93ek0f",
        B2u0y6b: "f1cgdal8",
        smv486: "f2b5gy5",
        u6pn5x: [
            "fk27hbk",
            "f1xelbol"
        ],
        B4uzyy: "f1gw7v5p",
        Be2s5ez: [
            "f1xelbol",
            "fk27hbk"
        ],
        xrcqlc: "f196ehyi",
        Bhxzhr1: [
            "frlu5y9",
            "f109kcev"
        ]
    },
    legendContainer: {
        xawz: 0,
        Bh6795r: 0,
        Bnnss6s: 0,
        fkmc3a: "fd43uyw",
        jrapky: 0,
        Frg6f3: 0,
        t21cq0: 0,
        B6of3ja: 0,
        B74szlk: "fduql59"
    },
    annotation: {
        mc9l5x: "f22iagw",
        Bt984gj: "f122n59",
        i8kkvl: 0,
        Belr9w4: 0,
        rmohyg: "f4xv25i"
    }
}, {
    d: [
        ".fz5stix{white-space:nowrap;}",
        ".fly5x3f{width:100%;}",
        ".f122n59{align-items:center;}",
        ".fpzxe1g{margin-top:-8px 0 0 -8px;}",
        ".f705vk6{margin-right:-8px 0 0 -8px;}",
        ".fqom4eq{margin-left:-8px 0 0 -8px;}",
        ".fzpzak3{margin-bottom:-8px 0 0 -8px;}",
        ".fc7pyfw{justify-content:left;}",
        ".fgbriwd{justify-content:right;}",
        ".f1k6fduh{cursor:pointer;}",
        ".f1ern45e{border-top-style:none;}",
        ".f1n71otn{border-right-style:none;}",
        ".f1deefiw{border-left-style:none;}",
        ".f1h8hb77{border-bottom-style:none;}",
        [
            ".f1t756mo{padding:8px;}",
            {
                p: -1
            }
        ],
        ".fiwaqkr{text-transform:capitalize;}",
        ".fy77jfu{min-width:0;}",
        ".frx94fk{width:12px;}",
        ".f1phki43{margin-right:8px;}",
        ".ff9s3yw{margin-left:8px;}",
        [
            ".fpyyuzi{border:1px solid;}",
            {
                p: -2
            }
        ],
        ".f3tsq5r{width:0;}",
        ".fniina8{height:0;}",
        ".fk1dseo{border-left-width:6px solid transparent;}",
        ".f1r9b9o{border-right-width:6px solid transparent;}",
        ".f1cm5upi{border-top-width:10.4px solid;}",
        ".f1vcna3q{margin-right:var(--spacingHorizontalS);}",
        ".foyynoy{margin-left:var(--spacingHorizontalS);}",
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fy9rknc{font-size:var(--fontSizeBase200);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".fwrc4pm{line-height:var(--lineHeightBase200);}",
        ".f19n0e5{color:var(--colorNeutralForeground1);}",
        ".fhv2zbx{forced-color-adjust:auto;}",
        ".fvblgha{height:12px;}",
        ".f1d7653x{border-top-width:1px solid;}",
        ".f1maxq7k{border-right-width:1px solid;}",
        ".f16g4l54{border-left-width:1px solid;}",
        ".f1rzi0e7{border-bottom-width:1px solid;}",
        ".f10pi13n{position:relative;}",
        ".f1o700av{text-align:left;}",
        ".fes3tcz{text-align:right;}",
        ".f15ivbov{transform:translate(-50%, 0);}",
        ".fjwp9lj{transform:translate(50%, 0);}",
        ".f198g47y{top:auto;}",
        ".f1ktbn1t{left:50%;}",
        ".ffenbu1{right:50%;}",
        ".f93ek0f{min-width:200px;}",
        ".f1cgdal8{max-width:800px;}",
        ".f2b5gy5::after{padding-top:1px 4px 1px;}",
        ".fk27hbk::after{padding-right:1px 4px 1px;}",
        ".f1xelbol::after{padding-left:1px 4px 1px;}",
        ".f1gw7v5p::after{padding-bottom:1px 4px 1px;}",
        ".f196ehyi::after{border-top-width:-2px;}",
        ".frlu5y9::after{border-left-width:-2px;}",
        ".f109kcev::after{border-right-width:-2px;}",
        [
            ".fd43uyw{flex:0 1 auto;}",
            {
                p: -1
            }
        ],
        [
            ".fduql59{margin:4px;}",
            {
                p: -1
            }
        ],
        ".f22iagw{display:flex;}",
        [
            ".f4xv25i{gap:4px;}",
            {
                p: -1
            }
        ]
    ],
    m: [
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.fuhkvgp{color:WindowText;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ],
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f15mrrko{forced-color-adjust:none;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ],
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f1vu5irt:hover{color:HighlightText;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ],
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f1sdboh2:hover{forced-color-adjust:none;}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ],
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f81dvq6{content:var(--rect-content-high-contrast);}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ],
        [
            "@media screen and (-ms-high-contrast: active),screen and (forced-colors: active){.f1a000x3{opacity:var(--rect-opacity-high-contrast);}}",
            {
                m: "screen and (-ms-high-contrast: active), screen and (forced-colors: active)"
            }
        ]
    ]
});
const useLegendStyles = (props)=>{
    var _props_styles, _props_styles1, _props_styles2, _props_styles3, _props_styles4, _props_styles5, _props_styles6, _props_styles7, _props_styles8, _props_styles9;
    const { className } = props; // ToDo - width, barHeight is non enumerable. Need to be used inline.
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(legendClassNames.root, baseStyles.root, className, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
        legend: (0, _react.mergeClasses)(legendClassNames.legend, baseStyles.legend, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.legend),
        rect: (0, _react.mergeClasses)(legendClassNames.rect, baseStyles.rect, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.rect),
        shape: (0, _react.mergeClasses)(legendClassNames.shape, baseStyles.shape, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.shape),
        triangle: (0, _react.mergeClasses)(legendClassNames.triangle, baseStyles.triangle, (_props_styles4 = props.styles) === null || _props_styles4 === void 0 ? void 0 : _props_styles4.triangle),
        text: (0, _react.mergeClasses)(legendClassNames.text, baseStyles.text, (_props_styles5 = props.styles) === null || _props_styles5 === void 0 ? void 0 : _props_styles5.text),
        hoverChange: (0, _react.mergeClasses)(legendClassNames.hoverChange, baseStyles.hoverChange, (_props_styles6 = props.styles) === null || _props_styles6 === void 0 ? void 0 : _props_styles6.hoverChange),
        resizableArea: (0, _react.mergeClasses)(legendClassNames.resizableArea, baseStyles.resizableArea, (_props_styles7 = props.styles) === null || _props_styles7 === void 0 ? void 0 : _props_styles7.resizableArea),
        legendContainer: (0, _react.mergeClasses)(legendClassNames.legendContainer, baseStyles.legendContainer, (_props_styles8 = props.styles) === null || _props_styles8 === void 0 ? void 0 : _props_styles8.legendContainer),
        annotation: (0, _react.mergeClasses)(legendClassNames.annotation, baseStyles.annotation, (_props_styles9 = props.styles) === null || _props_styles9 === void 0 ? void 0 : _props_styles9.annotation)
    };
};
