// @fluentui/react-charts/lib-commonjs/components/HeatMapChart/useHeatMapChartStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    heatmapChartClassNames: function() {
        return heatmapChartClassNames;
    },
    useHeatMapChartStyles: function() {
        return useHeatMapChartStyles;
    }
});
const _react = require("@griffel/react");
const heatmapChartClassNames = {
    root: 'fui-hmc__root',
    text: 'fui-hmc__text',
    calloutContentRoot: 'fui-hmc__calloutContentRoot',
    xAxis: '',
    yAxis: '',
    legendContainer: '',
    hover: '',
    descriptionMessage: '',
    tooltip: '',
    axisTitle: '',
    chartTitle: '',
    opacityChangeOnHover: '',
    shapeStyles: '',
    chartWrapper: '',
    svgTooltip: '',
    chart: '',
    axisAnnotation: ''
};
const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {},
    text: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fkhj508",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "f1i3iumi",
        Bkecrkj: "f1aehjj5"
    },
    calloutContentRoot: {
        B2u0y6b: "f15g8at8"
    }
}, {
    d: [
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fkhj508{font-size:var(--fontSizeBase300);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".f1i3iumi{line-height:var(--lineHeightBase300);}",
        ".f1aehjj5{pointer-events:none;}",
        ".f15g8at8{max-width:238px;}"
    ]
});
const useHeatMapChartStyles = (props)=>{
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(heatmapChartClassNames.root, baseStyles.root /*, props.styles?.root*/ ),
        text: (0, _react.mergeClasses)(heatmapChartClassNames.text, baseStyles.text /*, props.styles?.text*/ ),
        calloutContentRoot: (0, _react.mergeClasses)(heatmapChartClassNames.calloutContentRoot, baseStyles.calloutContentRoot /*, props.styles?.calloutContentRoot*/ )
    };
};
