// @fluentui/react-charts/lib/components/GaugeChart/useGaugeChartStyles.styles.js
'use client';

import { tokens, typographyStyles } from '@fluentui/react-theme';
import { __styles, mergeClasses } from '@griffel/react';
export const gaugeChartClassNames = {
  root: 'fui-gc__root',
  chart: 'fui-gc__chart',
  limits: 'fui-gc__limits',
  chartValue: 'fui-gc__chartValue',
  sublabel: 'fui-gc__sublabel',
  needle: 'fui-gc__needle',
  chartTitle: 'fui-gc__chartTitle',
  segment: 'fui-gc__segment',
  gradientSegment: 'fui-gc__gradientSegment',
  calloutContentRoot: 'fui-gc__calloutContentRoot',
  calloutDateTimeContainer: 'fui-gc__calloutDateTimeContainer',
  calloutContentX: 'fui-gc__calloutContentX',
  calloutBlockContainer: 'fui-gc__calloutBlockContainer',
  shapeStyles: 'fui-gc__shapeStyles',
  calloutlegendText: 'fui-gc__calloutlegendText',
  calloutContentY: 'fui-gc__calloutContentY',
  descriptionMessage: 'fui-gc__descriptionMessage',
  calloutInfoContainer: '',
  legendsContainer: 'fui-gc__legendsContainer',
  chartWrapper: 'fui-gc__chartWrapper'
};
const useStyles = /*#__PURE__*/__styles({
  root: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fkhj508",
    Bhrd7zp: "figsok6",
    Bg96gwp: "f1i3iumi",
    mc9l5x: "f22iagw",
    Beiy3e4: "f1vx9l62",
    Bt984gj: "f122n59",
    a9b677: "fly5x3f",
    Bqenvij: "f1l02sjl"
  },
  chart: {
    mc9l5x: "ftgm304"
  },
  limits: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "fwrc4pm",
    Bkfmm31: "fhuob2q",
    Bvjb7m6: "fhv2zbx"
  },
  chartValue: {
    Bhrd7zp: "fl43uef",
    Bkfmm31: "fhuob2q",
    Bvjb7m6: "fhv2zbx"
  },
  sublabel: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "fwrc4pm",
    Bkfmm31: "fhuob2q",
    Bvjb7m6: "fhv2zbx"
  },
  needle: {
    Bkfmm31: "fhuob2q",
    ojy3ng: "f1yuyku4"
  },
  chartTitle: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "figsok6",
    Bg96gwp: "fwrc4pm",
    Bkfmm31: "fhuob2q",
    Bvjb7m6: "fhv2zbx"
  },
  segment: {
    Bw0xxkn: 0,
    oeaueh: 0,
    Bpd4iqm: 0,
    Befb4lg: "f13htf1t",
    ojy3ng: "f1v24fw0"
  },
  gradientSegment: {
    a9b677: "fly5x3f",
    Bqenvij: "f1l02sjl"
  },
  calloutContentRoot: {
    mc9l5x: "f13qh94s",
    B68tc82: 0,
    Bmxbyg5: 0,
    Bpg54ce: "f1a3p1vp",
    De3pzq: "fxugw4r",
    jy2i9i: "fronft5"
  },
  calloutDateTimeContainer: {
    mc9l5x: "f22iagw",
    Beiy3e4: "f1063pyq",
    Brf1p80: "f1869bpl"
  },
  calloutContentX: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "figsok6",
    Bg96gwp: "f104wqfl",
    abs64n: "fd43bvn"
  },
  calloutBlockContainer: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fkhj508",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "f1i3iumi",
    B6of3ja: "f7jvbu2",
    sj55zd: "f19n0e5",
    uwmqm3: ["f177v4lu", "f19lj068"],
    mc9l5x: "ftgm304",
    Bvjb7m6: "fdgv6k0"
  },
  shapeStyles: {
    t21cq0: ["f1phki43", "ff9s3yw"]
  },
  calloutlegendText: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "figsok6",
    Bg96gwp: "f104wqfl",
    sj55zd: "fkfq4zb",
    Bvjb7m6: "fhv2zbx"
  },
  calloutContentY: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fkhj508",
    Bhrd7zp: "fl43uef",
    Bg96gwp: "f82itaf",
    Bvjb7m6: "fhv2zbx"
  },
  descriptionMessage: {
    Bahqtrf: "fk6fouc",
    Be2twd7: "fy9rknc",
    Bhrd7zp: "figsok6",
    Bg96gwp: "fwrc4pm",
    sj55zd: "f19n0e5",
    B6of3ja: "ffemd8d",
    z8tnut: "f1809wu7",
    g2u3we: 0,
    icvyot: 0,
    B4j52fo: 0,
    i8vvqc: "f10ci6s2"
  },
  legendsContainer: {
    a9b677: "fly5x3f"
  }
}, {
  d: [".fk6fouc{font-family:var(--fontFamilyBase);}", ".fkhj508{font-size:var(--fontSizeBase300);}", ".figsok6{font-weight:var(--fontWeightRegular);}", ".f1i3iumi{line-height:var(--lineHeightBase300);}", ".f22iagw{display:flex;}", ".f1vx9l62{flex-direction:column;}", ".f122n59{align-items:center;}", ".fly5x3f{width:100%;}", ".f1l02sjl{height:100%;}", ".ftgm304{display:block;}", ".fy9rknc{font-size:var(--fontSizeBase200);}", ".fl43uef{font-weight:var(--fontWeightSemibold);}", ".fwrc4pm{line-height:var(--lineHeightBase200);}", ".fhuob2q{fill:var(--colorNeutralForeground1);}", ".fhv2zbx{forced-color-adjust:auto;}", ".f1yuyku4{stroke:var(--colorNeutralBackground1);}", [".f13htf1t{outline:none;}", {
    p: -1
  }], ".f1v24fw0{stroke:var(--colorNeutralStroke1);}", ".f13qh94s{display:grid;}", [".f1a3p1vp{overflow:hidden;}", {
    p: -1
  }], ".fxugw4r{background-color:var(--colorNeutralBackground1);}", ".fronft5{background-blend-mode:normal,luminosity;}", ".f1063pyq{flex-direction:row;}", ".f1869bpl{justify-content:space-between;}", ".f104wqfl{line-height:16px;}", ".fd43bvn{opacity:0.85;}", ".f7jvbu2{margin-top:13px;}", ".f19n0e5{color:var(--colorNeutralForeground1);}", ".f177v4lu{padding-left:8px;}", ".f19lj068{padding-right:8px;}", ".fdgv6k0{forced-color-adjust:none;}", ".f1phki43{margin-right:8px;}", ".ff9s3yw{margin-left:8px;}", ".fkfq4zb{color:var(--colorNeutralForeground2);}", ".f82itaf{line-height:22px;}", ".ffemd8d{margin-top:10px;}", ".f1809wu7{padding-top:10px;}", [".f10ci6s2{border-top:1px solid var(--colorNeutralStroke1);}", {
    p: -1
  }]]
});
export const useGaugeChartStyles = props => {
  var _props_styles, _props_styles1, _props_styles2, _props_styles3, _props_styles4, _props_styles5, _props_styles6, _props_styles7, _props_styles8, _props_styles9, _props_styles10, _props_styles11, _props_styles12, _props_styles13, _props_styles14, _props_styles15, _props_styles16, _props_styles17, _props_styles18;
  const baseStyles = useStyles();
  return {
    root: mergeClasses(gaugeChartClassNames.root, baseStyles.root, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.root),
    chart: mergeClasses(gaugeChartClassNames.chart, baseStyles.chart, (_props_styles1 = props.styles) === null || _props_styles1 === void 0 ? void 0 : _props_styles1.chart),
    limits: mergeClasses(gaugeChartClassNames.limits, baseStyles.limits, (_props_styles2 = props.styles) === null || _props_styles2 === void 0 ? void 0 : _props_styles2.limits),
    chartValue: mergeClasses(gaugeChartClassNames.chartValue, baseStyles.chartValue, (_props_styles3 = props.styles) === null || _props_styles3 === void 0 ? void 0 : _props_styles3.chartValue),
    sublabel: mergeClasses(gaugeChartClassNames.sublabel, baseStyles.sublabel, (_props_styles4 = props.styles) === null || _props_styles4 === void 0 ? void 0 : _props_styles4.sublabel),
    needle: mergeClasses(gaugeChartClassNames.needle, baseStyles.needle, (_props_styles5 = props.styles) === null || _props_styles5 === void 0 ? void 0 : _props_styles5.needle),
    chartTitle: mergeClasses(gaugeChartClassNames.chartTitle, baseStyles.chartTitle, (_props_styles6 = props.styles) === null || _props_styles6 === void 0 ? void 0 : _props_styles6.chartTitle),
    segment: mergeClasses(gaugeChartClassNames.segment, baseStyles.segment, (_props_styles7 = props.styles) === null || _props_styles7 === void 0 ? void 0 : _props_styles7.segment),
    gradientSegment: mergeClasses(gaugeChartClassNames.gradientSegment, baseStyles.gradientSegment, (_props_styles8 = props.styles) === null || _props_styles8 === void 0 ? void 0 : _props_styles8.gradientSegment),
    calloutContentRoot: mergeClasses(gaugeChartClassNames.calloutContentRoot, baseStyles.calloutContentRoot, (_props_styles9 = props.styles) === null || _props_styles9 === void 0 ? void 0 : _props_styles9.calloutContentRoot),
    calloutDateTimeContainer: mergeClasses(gaugeChartClassNames.calloutDateTimeContainer, baseStyles.calloutDateTimeContainer, (_props_styles10 = props.styles) === null || _props_styles10 === void 0 ? void 0 : _props_styles10.calloutDateTimeContainer),
    calloutContentX: mergeClasses(gaugeChartClassNames.calloutContentX, baseStyles.calloutContentX, (_props_styles11 = props.styles) === null || _props_styles11 === void 0 ? void 0 : _props_styles11.calloutContentX),
    calloutBlockContainer: mergeClasses(gaugeChartClassNames.calloutBlockContainer, baseStyles.calloutBlockContainer, (_props_styles12 = props.styles) === null || _props_styles12 === void 0 ? void 0 : _props_styles12.calloutBlockContainer),
    shapeStyles: mergeClasses(gaugeChartClassNames.shapeStyles, baseStyles.shapeStyles, (_props_styles13 = props.styles) === null || _props_styles13 === void 0 ? void 0 : _props_styles13.shapeStyles),
    calloutlegendText: mergeClasses(gaugeChartClassNames.calloutlegendText, baseStyles.calloutlegendText, (_props_styles14 = props.styles) === null || _props_styles14 === void 0 ? void 0 : _props_styles14.calloutlegendText),
    calloutContentY: mergeClasses(gaugeChartClassNames.calloutContentY, baseStyles.calloutContentY, (_props_styles15 = props.styles) === null || _props_styles15 === void 0 ? void 0 : _props_styles15.calloutContentY),
    descriptionMessage: mergeClasses(gaugeChartClassNames.descriptionMessage, baseStyles.descriptionMessage, (_props_styles16 = props.styles) === null || _props_styles16 === void 0 ? void 0 : _props_styles16.descriptionMessage),
    chartWrapper: mergeClasses(gaugeChartClassNames.chartWrapper, (_props_styles17 = props.styles) === null || _props_styles17 === void 0 ? void 0 : _props_styles17.chartWrapper),
    legendsContainer: mergeClasses(gaugeChartClassNames.legendsContainer, baseStyles.legendsContainer, (_props_styles18 = props.styles) === null || _props_styles18 === void 0 ? void 0 : _props_styles18.legendsContainer)
  };
};
