// @fluentui/react-charts/lib-commonjs/components/ResponsiveContainer/useResponsiveChildStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    responsiveChildClassNames: function() {
        return responsiveChildClassNames;
    },
    useResponsiveChildStyles: function() {
        return useResponsiveChildStyles;
    }
});
const _react = require("@griffel/react");
const responsiveChildClassNames = {
    root: 'fui-charts-resp-child__root',
    chartWrapper: 'fui-charts-resp-child__chartWrapper',
    chart: 'fui-charts-resp-child__chart'
};
const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    root: {
        a9b677: "fly5x3f",
        Bqenvij: "f1l02sjl"
    },
    chartWrapper: {
        a9b677: "fly5x3f"
    },
    chart: {
        a9b677: "fly5x3f"
    }
}, {
    d: [
        ".fly5x3f{width:100%;}",
        ".f1l02sjl{height:100%;}"
    ]
});
const useResponsiveChildStyles = ()=>{
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(responsiveChildClassNames.root, baseStyles.root),
        chartWrapper: (0, _react.mergeClasses)(responsiveChildClassNames.chartWrapper, baseStyles.chartWrapper),
        chart: (0, _react.mergeClasses)(responsiveChildClassNames.chart, baseStyles.chart)
    };
};
