// @fluentui/react-charts/lib-commonjs/components/ResponsiveContainer/useResponsiveChildStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    responsiveChildClassNames: function() {
        return responsiveChildClassNames;
    },
    useResponsiveChildStyles: function() {
        return useResponsiveChildStyles;
    }
});
const _react = require("@griffel/react");
const responsiveChildClassNames = {
    root: 'fui-charts-resp-child__root',
    chartWrapper: 'fui-charts-resp-child__chartWrapper',
    chart: 'fui-charts-resp-child__chart'
};
const useStyles = (0, _react.makeStyles)({
    root: {
        width: '100%',
        height: '100%'
    },
    chartWrapper: {
        width: '100%'
    },
    chart: {
        // This overrides the pixel width of svg allowing it to resize properly within a flexbox or grid layout.
        // Note: height is not set to 100% because that causes the charts to resize vertically in an infinite loop.
        width: '100%'
    }
});
const useResponsiveChildStyles = ()=>{
    const baseStyles = useStyles();
    return {
        root: (0, _react.mergeClasses)(responsiveChildClassNames.root, baseStyles.root),
        chartWrapper: (0, _react.mergeClasses)(responsiveChildClassNames.chartWrapper, baseStyles.chartWrapper),
        chart: (0, _react.mergeClasses)(responsiveChildClassNames.chart, baseStyles.chart)
    };
};
