// @fluentui/react-charts/lib-commonjs/components/ResponsiveContainer/withResponsiveContainer.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "withResponsiveContainer", {
    enumerable: true,
    get: function() {
        return withResponsiveContainer;
    }
});
const _interop_require_wildcard = require("@swc/helpers/_/_interop_require_wildcard");
const _react = /*#__PURE__*/ _interop_require_wildcard._(require("react"));
const _ResponsiveContainer = require("./ResponsiveContainer");
function withResponsiveContainer(WrappedComponent) {
    const ComponentWithResponsiveContainer = ({ aspect, width, height, minWidth, minHeight, maxHeight, onResize, ...restProps })=>{
        return /*#__PURE__*/ _react.createElement(_ResponsiveContainer.ResponsiveContainer, {
            aspect: aspect,
            width: width,
            height: height,
            minWidth: minWidth,
            minHeight: minHeight,
            maxHeight: maxHeight,
            onResize: onResize
        }, /*#__PURE__*/ _react.createElement(WrappedComponent, restProps));
    };
    return ComponentWithResponsiveContainer;
}
