// @fluentui/react-charts/lib-commonjs/components/CommonComponents/useChartPopoverStyles.styles.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    popoverClassNames: function() {
        return popoverClassNames;
    },
    usePopoverStyles_unstable: function() {
        return usePopoverStyles_unstable;
    }
});
const _react = require("@griffel/react");
const popoverClassNames = {
    calloutContentRoot: 'fui-cart__calloutContentRoot',
    calloutDateTimeContainer: 'fui-cart__calloutDateTimeContainer',
    calloutContentX: 'fui-cart__calloutContentX',
    calloutBlockContainer: 'fui-cart__calloutBlockContainer',
    calloutBlockContainertoDrawShapefalse: 'fui-cart__calloutBlockContainertoDrawShapefalse',
    calloutBlockContainertoDrawShapetrue: 'fui-cart__calloutBlockContainertoDrawShapetrue',
    shapeStyles: 'fui-cart__shapeStyles',
    calloutlegendText: 'fui-cart__calloutlegendText',
    calloutContentY: 'fui-cart__calloutContentY',
    descriptionMessage: 'fui-cart__descriptionMessage',
    ratio: 'fui-cart__ratio',
    numerator: 'fui-cart__numerator',
    denominator: 'fui-cart__denominator',
    calloutInfoContainer: 'fui-cart__calloutInfoContainer',
    calloutContainer: 'fui-cart__calloutContainer'
};
/**
 * Base Styles
 */ const useStyles = /*#__PURE__*/ (0, _react.__styles)({
    calloutContentRoot: {
        mc9l5x: "f13qh94s",
        B68tc82: 0,
        Bmxbyg5: 0,
        Bpg54ce: "f1a3p1vp",
        z8tnut: "f1dvrhzg",
        z189sj: [
            "fjnv7mu",
            "fae8789"
        ],
        Byoj8tv: "f1av7aef",
        uwmqm3: [
            "fae8789",
            "fjnv7mu"
        ],
        De3pzq: "fxugw4r",
        jy2i9i: "fronft5"
    },
    calloutDateTimeContainer: {
        mc9l5x: "f22iagw",
        Beiy3e4: "f1063pyq",
        Brf1p80: "f1869bpl"
    },
    calloutContentX: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fwrc4pm",
        abs64n: "fp25eh",
        sj55zd: "fkfq4zb"
    },
    calloutBlockContainer: {
        sj55zd: "fkfq4zb"
    },
    calloutBlockContainerCartesian: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fwrc4pm",
        Bvjb7m6: "fdgv6k0"
    },
    calloutBlockContainerNonCartesian: {
        Be2twd7: "fojgt09",
        Bg96gwp: "f82itaf",
        Bvjb7m6: "fdgv6k0"
    },
    calloutBlockContainertoDrawShapefalse: {
        uwmqm3: [
            "f1f5gg8d",
            "f1vdfbxk"
        ],
        Bvjb7m6: "fdgv6k0"
    },
    calloutBlockContainertoDrawShapetrue: {
        mc9l5x: "fwk3njj"
    },
    shapeStyles: {
        t21cq0: [
            "f1vcna3q",
            "foyynoy"
        ]
    },
    calloutLegendText: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fwrc4pm",
        sj55zd: "fkfq4zb",
        Bvjb7m6: "fhv2zbx",
        jrapky: "f8l5zjj"
    },
    calloutContentY: {
        Bvjb7m6: "fhv2zbx"
    },
    calloutContentYCartesian: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fod5ikn",
        Bhrd7zp: "flh3ekv",
        Bg96gwp: "faaz57k"
    },
    calloutContentYNonCartesian: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fojgt09",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fcen8rp"
    },
    descriptionMessage: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "fy9rknc",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fwrc4pm",
        sj55zd: "fkfq4zb",
        B6of3ja: "f1gaxbfw",
        z8tnut: "f1ngh7ph",
        g2u3we: 0,
        icvyot: 0,
        B4j52fo: 0,
        i8vvqc: "f16zin0e"
    },
    ratio: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "f13mqy1h",
        Bhrd7zp: "figsok6",
        Bg96gwp: "fcpl73t",
        Frg6f3: [
            "f1t5qyk5",
            "f1ikr372"
        ],
        sj55zd: "f19n0e5"
    },
    numerator: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "f13mqy1h",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fcpl73t"
    },
    denominator: {
        Bahqtrf: "fk6fouc",
        Be2twd7: "f13mqy1h",
        Bhrd7zp: "fl43uef",
        Bg96gwp: "fcpl73t"
    },
    calloutInfoContainer: {
        uwmqm3: [
            "f1f5gg8d",
            "f1vdfbxk"
        ],
        Bvjb7m6: "fdgv6k0"
    },
    calloutContainer: {}
}, {
    d: [
        ".f13qh94s{display:grid;}",
        [
            ".f1a3p1vp{overflow:hidden;}",
            {
                p: -1
            }
        ],
        ".f1dvrhzg{padding-top:11px 16px 10px 16px;}",
        ".fjnv7mu{padding-right:11px 16px 10px 16px;}",
        ".fae8789{padding-left:11px 16px 10px 16px;}",
        ".f1av7aef{padding-bottom:11px 16px 10px 16px;}",
        ".fxugw4r{background-color:var(--colorNeutralBackground1);}",
        ".fronft5{background-blend-mode:normal,luminosity;}",
        ".f22iagw{display:flex;}",
        ".f1063pyq{flex-direction:row;}",
        ".f1869bpl{justify-content:space-between;}",
        ".fk6fouc{font-family:var(--fontFamilyBase);}",
        ".fy9rknc{font-size:var(--fontSizeBase200);}",
        ".figsok6{font-weight:var(--fontWeightRegular);}",
        ".fwrc4pm{line-height:var(--lineHeightBase200);}",
        ".fp25eh{opacity:0.8;}",
        ".fkfq4zb{color:var(--colorNeutralForeground2);}",
        ".fdgv6k0{forced-color-adjust:none;}",
        ".fojgt09{font-size:var(--fontSizeHero700);}",
        ".f82itaf{line-height:22px;}",
        ".f1f5gg8d{padding-left:var(--spacingHorizontalS);}",
        ".f1vdfbxk{padding-right:var(--spacingHorizontalS);}",
        ".fwk3njj{display:inline-grid;}",
        ".f1vcna3q{margin-right:var(--spacingHorizontalS);}",
        ".foyynoy{margin-left:var(--spacingHorizontalS);}",
        ".fhv2zbx{forced-color-adjust:auto;}",
        ".f8l5zjj{margin-bottom:var(--spacingVerticalXS);}",
        ".fod5ikn{font-size:var(--fontSizeBase400);}",
        ".flh3ekv{font-weight:var(--fontWeightBold);}",
        ".faaz57k{line-height:var(--lineHeightBase400);}",
        ".fl43uef{font-weight:var(--fontWeightSemibold);}",
        ".fcen8rp{line-height:var(--lineHeightHero700);}",
        ".f1gaxbfw{margin-top:var(--spacingVerticalMNudge);}",
        ".f1ngh7ph{padding-top:var(--spacingVerticalMNudge);}",
        [
            ".f16zin0e{border-top:1px solid var(--colorNeutralStroke2);}",
            {
                p: -1
            }
        ],
        ".f13mqy1h{font-size:var(--fontSizeBase100);}",
        ".fcpl73t{line-height:var(--lineHeightBase100);}",
        ".f1t5qyk5{margin-left:var(--spacingHorizontalSNudge);}",
        ".f1ikr372{margin-right:var(--spacingHorizontalSNudge);}",
        ".f19n0e5{color:var(--colorNeutralForeground1);}"
    ]
});
const usePopoverStyles_unstable = (props)=>{
    var _props_styles;
    const { isCartesian } = props;
    const baseStyles = useStyles();
    return {
        calloutContentRoot: (0, _react.mergeClasses)(popoverClassNames.calloutContentRoot, baseStyles.calloutContentRoot, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.calloutContentRoot),
        calloutDateTimeContainer: (0, _react.mergeClasses)(popoverClassNames.calloutDateTimeContainer, baseStyles.calloutDateTimeContainer /*props.styles?.calloutDateTimeContainer*/ ),
        calloutContentX: (0, _react.mergeClasses)(popoverClassNames.calloutContentX, baseStyles.calloutContentX /*props.styles?.calloutContentX*/ ),
        calloutBlockContainer: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainer, baseStyles.calloutBlockContainer /*props.styles?.calloutBlockContainerCartesian*/ , isCartesian ? baseStyles.calloutBlockContainerCartesian : baseStyles.calloutBlockContainerNonCartesian),
        calloutBlockContainertoDrawShapefalse: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainertoDrawShapefalse, baseStyles.calloutBlockContainertoDrawShapefalse /*props.styles?.calloutBlockContainertoDrawShapefalse*/ ),
        calloutBlockContainertoDrawShapetrue: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainertoDrawShapetrue, baseStyles.calloutBlockContainertoDrawShapetrue /*props.styles?.calloutBlockContainertoDrawShapetrue*/ ),
        shapeStyles: (0, _react.mergeClasses)(popoverClassNames.shapeStyles, baseStyles.shapeStyles /*props.styles?.shapeStyles*/ ),
        calloutlegendText: (0, _react.mergeClasses)(popoverClassNames.calloutlegendText, baseStyles.calloutLegendText /*props.styles?.calloutlegendText*/ ),
        calloutContentY: (0, _react.mergeClasses)(popoverClassNames.calloutContentY, baseStyles.calloutContentY /*props.styles?.calloutContentYNonCartesian*/ , isCartesian ? baseStyles.calloutContentYCartesian : baseStyles.calloutContentYNonCartesian),
        descriptionMessage: (0, _react.mergeClasses)(popoverClassNames.descriptionMessage, baseStyles.descriptionMessage /*props.styles?. descriptionMessage*/ ),
        ratio: (0, _react.mergeClasses)(popoverClassNames.ratio, baseStyles.ratio /*props.styles?.ratio*/ ),
        numerator: (0, _react.mergeClasses)(popoverClassNames.numerator, baseStyles.numerator /*props.styles?.numerator*/ ),
        denominator: (0, _react.mergeClasses)(popoverClassNames.denominator, baseStyles.denominator /*props.styles?.denominator*/ ),
        calloutInfoContainer: (0, _react.mergeClasses)(popoverClassNames.calloutInfoContainer, baseStyles.calloutInfoContainer),
        calloutContainer: (0, _react.mergeClasses)(popoverClassNames.calloutContainer, baseStyles.calloutContainer)
    };
};
