// @fluentui/react-charts/lib-commonjs/components/CommonComponents/useChartPopoverStyles.styles.raw.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    popoverClassNames: function() {
        return popoverClassNames;
    },
    usePopoverStyles_unstable: function() {
        return usePopoverStyles_unstable;
    }
});
const _react = require("@griffel/react");
const _reacttheme = require("@fluentui/react-theme");
const popoverClassNames = {
    calloutContentRoot: 'fui-cart__calloutContentRoot',
    calloutDateTimeContainer: 'fui-cart__calloutDateTimeContainer',
    calloutContentX: 'fui-cart__calloutContentX',
    calloutBlockContainer: 'fui-cart__calloutBlockContainer',
    calloutBlockContainertoDrawShapefalse: 'fui-cart__calloutBlockContainertoDrawShapefalse',
    calloutBlockContainertoDrawShapetrue: 'fui-cart__calloutBlockContainertoDrawShapetrue',
    shapeStyles: 'fui-cart__shapeStyles',
    calloutlegendText: 'fui-cart__calloutlegendText',
    calloutContentY: 'fui-cart__calloutContentY',
    descriptionMessage: 'fui-cart__descriptionMessage',
    ratio: 'fui-cart__ratio',
    numerator: 'fui-cart__numerator',
    denominator: 'fui-cart__denominator',
    calloutInfoContainer: 'fui-cart__calloutInfoContainer',
    calloutContainer: 'fui-cart__calloutContainer'
};
/**
 * Base Styles
 */ const useStyles = (0, _react.makeStyles)({
    calloutContentRoot: {
        display: 'grid',
        overflow: 'hidden',
        ..._react.shorthands.padding('11px 16px 10px 16px'),
        backgroundColor: _reacttheme.tokens.colorNeutralBackground1,
        backgroundBlendMode: 'normal, luminosity'
    },
    calloutDateTimeContainer: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    calloutContentX: {
        ..._reacttheme.typographyStyles.caption1,
        opacity: '0.8',
        color: _reacttheme.tokens.colorNeutralForeground2
    },
    calloutBlockContainer: {
        color: _reacttheme.tokens.colorNeutralForeground2
    },
    calloutBlockContainerCartesian: {
        ..._reacttheme.typographyStyles.caption1,
        forcedColorAdjust: 'none'
    },
    calloutBlockContainerNonCartesian: {
        fontSize: _reacttheme.tokens.fontSizeHero700,
        lineHeight: '22px',
        forcedColorAdjust: 'none'
    },
    calloutBlockContainertoDrawShapefalse: {
        paddingLeft: _reacttheme.tokens.spacingHorizontalS,
        forcedColorAdjust: 'none'
    },
    calloutBlockContainertoDrawShapetrue: {
        display: 'inline-grid'
    },
    shapeStyles: {
        marginRight: _reacttheme.tokens.spacingHorizontalS
    },
    calloutLegendText: {
        ..._reacttheme.typographyStyles.caption1,
        color: _reacttheme.tokens.colorNeutralForeground2,
        forcedColorAdjust: 'auto',
        marginBottom: _reacttheme.tokens.spacingVerticalXS
    },
    calloutContentY: {
        forcedColorAdjust: 'auto'
    },
    calloutContentYCartesian: {
        ..._reacttheme.typographyStyles.subtitle2Stronger
    },
    calloutContentYNonCartesian: {
        ..._reacttheme.typographyStyles.title2
    },
    descriptionMessage: {
        ..._reacttheme.typographyStyles.caption1,
        color: _reacttheme.tokens.colorNeutralForeground2,
        marginTop: _reacttheme.tokens.spacingVerticalMNudge,
        paddingTop: _reacttheme.tokens.spacingVerticalMNudge,
        borderTop: `1px solid ${_reacttheme.tokens.colorNeutralStroke2}`
    },
    ratio: {
        ..._reacttheme.typographyStyles.caption2,
        marginLeft: _reacttheme.tokens.spacingHorizontalSNudge,
        color: _reacttheme.tokens.colorNeutralForeground1
    },
    numerator: {
        ..._reacttheme.typographyStyles.caption2Strong
    },
    denominator: {
        ..._reacttheme.typographyStyles.caption2Strong
    },
    calloutInfoContainer: {
        paddingLeft: _reacttheme.tokens.spacingHorizontalS,
        forcedColorAdjust: 'none'
    },
    calloutContainer: {}
});
const usePopoverStyles_unstable = (props)=>{
    var _props_styles;
    const { isCartesian } = props;
    const baseStyles = useStyles();
    return {
        calloutContentRoot: (0, _react.mergeClasses)(popoverClassNames.calloutContentRoot, baseStyles.calloutContentRoot, (_props_styles = props.styles) === null || _props_styles === void 0 ? void 0 : _props_styles.calloutContentRoot),
        calloutDateTimeContainer: (0, _react.mergeClasses)(popoverClassNames.calloutDateTimeContainer, baseStyles.calloutDateTimeContainer /*props.styles?.calloutDateTimeContainer*/ ),
        calloutContentX: (0, _react.mergeClasses)(popoverClassNames.calloutContentX, baseStyles.calloutContentX /*props.styles?.calloutContentX*/ ),
        calloutBlockContainer: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainer, baseStyles.calloutBlockContainer /*props.styles?.calloutBlockContainerCartesian*/ , isCartesian ? baseStyles.calloutBlockContainerCartesian : baseStyles.calloutBlockContainerNonCartesian),
        calloutBlockContainertoDrawShapefalse: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainertoDrawShapefalse, baseStyles.calloutBlockContainertoDrawShapefalse /*props.styles?.calloutBlockContainertoDrawShapefalse*/ ),
        calloutBlockContainertoDrawShapetrue: (0, _react.mergeClasses)(popoverClassNames.calloutBlockContainertoDrawShapetrue, baseStyles.calloutBlockContainertoDrawShapetrue /*props.styles?.calloutBlockContainertoDrawShapetrue*/ ),
        shapeStyles: (0, _react.mergeClasses)(popoverClassNames.shapeStyles, baseStyles.shapeStyles /*props.styles?.shapeStyles*/ ),
        calloutlegendText: (0, _react.mergeClasses)(popoverClassNames.calloutlegendText, baseStyles.calloutLegendText /*props.styles?.calloutlegendText*/ ),
        calloutContentY: (0, _react.mergeClasses)(popoverClassNames.calloutContentY, baseStyles.calloutContentY /*props.styles?.calloutContentYNonCartesian*/ , isCartesian ? baseStyles.calloutContentYCartesian : baseStyles.calloutContentYNonCartesian),
        descriptionMessage: (0, _react.mergeClasses)(popoverClassNames.descriptionMessage, baseStyles.descriptionMessage /*props.styles?. descriptionMessage*/ ),
        ratio: (0, _react.mergeClasses)(popoverClassNames.ratio, baseStyles.ratio /*props.styles?.ratio*/ ),
        numerator: (0, _react.mergeClasses)(popoverClassNames.numerator, baseStyles.numerator /*props.styles?.numerator*/ ),
        denominator: (0, _react.mergeClasses)(popoverClassNames.denominator, baseStyles.denominator /*props.styles?.denominator*/ ),
        calloutInfoContainer: (0, _react.mergeClasses)(popoverClassNames.calloutInfoContainer, baseStyles.calloutInfoContainer),
        calloutContainer: (0, _react.mergeClasses)(popoverClassNames.calloutContainer, baseStyles.calloutContainer)
    };
};
